### ISSUE TITLE
proposal: onReady and onError for Connection

### ISSUE THREAD

**Issue by pabigot** [CopernicaMarketingSoftware/AMQP-CPP#220]
AMQP-CPP is meeting my needs very well, but I've noticed one gap in API that's complicating things a little:

Channel objects have onReady() methods that call back when you can use the channel, but Connection objects do not.  The documentation I see suggests that applications that don't want to rely on AMQP-CPP's deferred connection operations should extend ConnectionHandler to detect completion before operating on it.

Some of my uses are intermediaries which might take data from one broker, transform it, and send it to another.  If a connection to the downstream consumer doesn't succeed I don't want to have pending operations on the upstream producer that I need to roll back, even if they could be blocked by Channel::onReady() conditions, so the deferred operation approach is not ideal.  With multiple producers and consumers the reaction to a successful/failed connection may differ depending on the connection instance, so having to code this into a subclass of ConnectionHandler is inconvenient: at a minimum it would become dependent on the intermediate event loop handler implementation.

My proposal is to add two methods to Connection:

* onReady() gets invoked in ConnectionImpl::setConnected() after handler::onConnected() if the handler doesn't destroy the connection;

* onError(const char * error) gets invoked by ConnectionImpl::reportError after handler::onError if the handler doesn't destroy the connection.

Without implementing it I believe these would meet my needs without interfering with existing library uses.  I don't believe an onClose() is warranted since non-error close is under the control of the application.

If I prepare a patch that adds these, is there a potential for it to be merged?  Or alternative approaches/APIs that would be preferred?  Or am I missing some existing way to monitor critical connection state changes (viz. successful open and asynchronous error) without having to put that code into the event handler implementation or forward it through channel handlers?

**Comment by pabigot** [comment]
With implementing it seems this is completely wrong, as Connection isn't a superclass of TcpConnection which is what needs this capability in my applications.  TcpConnection could be extended to provide this sort of notification but it's a ConnectionHandler which already uses all these names with private implementations.  Additional thought is required.

**Comment by EmielBruijntjes** [comment]
To be honest, I do not really like the Channel::onSomething() methods. I prefer the setup in which the callbacks are attached to the specific operation (via de Deferred class), instead of to the channel as a whole. So the initiative to add such "global" callbacks to the Connection class too is not something that I like.

**Comment by pabigot** [comment]
That make sense, but since constructing a TcpConnection implicitly initiates the connection operation, there's nothing else to attach it to.  Agreed it can't go on Connection, but it's fairly easy to add onReady() and onFailed() methods to TcpConnection, with the TcpState infrastructure responsible for invoking the appropriate one of those at the point where the connection succeeds or fails.

While implementing that I found some areas where I think the current implementation might invoke a callback with a pointer to a destructed TcpConnection.

I've got something that appears to work; if it still seems like a good idea after I've used it for a while I'll build a PR.

**Comment by RZRCDR** [comment]
I have done something very similar although I am not using the build in TcpConnection Class.
I have created a derived class "connectionHandler" inheriting from AMQP:ConnectionHandler and based on Boost::Asio.

I am propagating the errors on the TCP connection up by means of a callback as you also suggested.
You register an error callback against the connection_handler_t, that gives you the ability to get notified about these lower layer errors and then tear down all your AMQP specifics and re-initiate the whole connection and AMQP protocol communication.

Here is the example.

[src.zip](https://github.com/CopernicaMarketingSoftware/AMQP-CPP/files/1946454/src.zip)
using error_cb = std::function<void(const std::string &error)>;   // error callback
connection_handler_t &set_on_error(const error_cb &cb);

**Comment by pabigot** [comment]
@EmielBruijntjes Thinking more deeply about your comment, I don't understand.  The existing Channel.onSomething() methods--specifically `onReady`, `onError`, and when PR #134 is merged `onAck` and `onNack`--notify about channel-level events for which there is no deferred operation result to which they could be attached.  How else would you get those events?

My primary unhappiness with the existing API is that those methods don't return a `Channel&` to allow chaining, which makes them inconsistent with all the `Deferred::onSomething` methods.

**Comment by EmielBruijntjes** [comment]
The Channel::onReady() and Channel::onError() methods are here because I could not think of a better way to do this. But I do not like them, because they pollute the very generic channel object with members that are only used in specific situations. That's also why I have still not merged #134.

**Comment by pabigot** [comment]
Well, ok.  I understand AMQP-CPP supports throwaway channels for one-shot operations, and that's useful but it's not how I generally use them.  In my applications waiting for ready before initiating a transaction and being notified of channel errors are not specific situations (especially since any error shuts down the channel permanently).  Ack/Nack for publisher confirms aren't either since they're not isolated to (and in fact are asynchronous with) a specific invocation of publish.

Nor do I have an alternative solution, except the way I've had to go with the TcpConnection which is to move the notification infrastructure to a subclass of ConnectionHandler which maintains a per-connection registry of callbacks---which is unworkable for Channels because unlike TcpConnection the ConnectionHandler isn't notified of changes to channel status.

I suppose we could come up with a ChannelHandler object, but IMO it'd just be an annoying intermediary since the recovery of error on a specific channel instance will be specific to that instance.

So how do we resolve this?  I need #134; if it's not going to get merged soon then I'll need to start maintaining a forked AMPQ-CPP.

**Comment by EmielBruijntjes** [comment]
@pabigot Is this issue still alive? Pull request #134 has been merged, but using an alternative interface.

**Comment by pabigot** [comment]
No, the solution in #134 worked for me.
